training/heterogeneous-clusters/pt.grpc.sagemaker/hetero-pytorch-mnist.ipynb

```diff
@@ -0,0 +1,509 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "id": "85da9619",
+   "metadata": {},
+   "source": [
+    "## PyTorch's example to demonstrate Amazon SageMaker Heterogeneous Cluster for model training\n",
+    "\n",
+    "---\n",
+    "### Description\n",
+    "Heterogeneous clusters enable launching training jobs that use multiple instance types in a single job. This capability can improve your training cost and speed by running different parts of the model training on the most suitable instance type. This use case typically happens in computer vision DL training, where training is bottleneck on CPU resources needed for data augmentation, leaving the expensive GPU underutilized. Heterogeneous clusters enable you to add more CPU resources to fully utilize GPUs, thus increase training speed and cost-efficiency. For more details, you can find the documentation of this feature [here](https://docs.aws.amazon.com/sagemaker/latest/dg/train-heterogeneous-cluster.html).\n",
+    "\n",
+    "This notebook demonstrates how to use Heterogeneous Cluster feature of SageMaker Training with PyTorch 1.10.  The notebook works on Python 3 (_PyTorch 1.10 Python 3.8 CPU Optimized_) image of SageMaker Studio Notebook instance, and runs on _ml.t3.medium_ instance type.\n",
+    "\n",
+    "In this sample notebook, we have taken the PyTorch model based on this [official MNIST example](https://github.com/pytorch/examples/tree/main/MNIST). We modified the training code to be heavy on data pre-processing. We are going to train this model in both Homogeneous and Heterogeneous Cluster modes. The flag to train on any of these modes can be set using `IS_HETERO = False or True` in section **B.2 Configure environment variables**. \n",
+    "\n",
+    "<table>\n",
+    "  <tr>\n",
+    "    <td width=43%><b>Homogeneous training job</b><br/>\n",
+    "    We first run the homogeneous training job to get baseline performance number, and observe that it is unable to fully utilize GPU of ml.p3.2xlarge instance due to a CPU being the bottleneck.</td>\n",
```

Comment:
These instance types are not always available across the major commercial regions.